test/rational.jl

```diff
@@ -745,3 +745,318 @@ end
     @test rationalize(0.1im; tol=eps(0.1)) == rationalize(0.1im)
 end
+
+naive_rational_to_bigfloat(x) = BigFloat(numerator(x))/BigFloat(denominator(x))
+
+function test_rational_to_float_systematic_with_naive_impl(::Type{F}, q, rm) where {F}
+    no_alloc_rat = let
+        f(::Rational{Bool}) = true
+        function f(x)
+            if (sizeof(Int) < 8) && (8 < sizeof(x))
+                # `div(::Int128, ::Int128)` is implemented via `BigInt`on 32-bit Julia
+                false
+            else
+                isbits(widen(x))
+            end
+        end
+        f
+    end
+    want_no_alloc = no_alloc_rat(q) && isbitstype(widen(F))
+    want_no_alloc && @test iszero(@allocated F(q, rm))
+    @test F(naive_rational_to_bigfloat(q), rm) == @inferred F(q, rm)
+end
+
+function test_rational_to_float_systematic_with_naive_impl(::Type{BigFloat}, q, rm, prec)
+    reference = BigFloat(naive_rational_to_bigfloat(q), rm, precision = prec)
+    @test precision(reference) == prec
+    t = let reference = reference, prec = prec
+        function(res)
+            @test res == reference
+            @test precision(res) == prec
+        end
+    end
+
+    t(BigFloat(q, rm, precision = prec))
+
+    let
+        res = setprecision(BigFloat, prec) do
+            BigFloat(q, rm)
+        end
+        t(res)
+    end
+
+    let
+        if rm isa RoundingMode
+            res = setrounding(BigFloat, rm) do
+                BigFloat(q, precision = prec)
+            end
+            t(res)
+        end
+    end
+end
+
+function test_rational_to_float_systematic_with_naive_impl(::Type{BigFloat}, q, rm)
+    for prec ∈ Base.OneTo(10)
+        test_rational_to_float_systematic_with_naive_impl(BigFloat, q, rm, prec)
+    end
+end
+
+function test_rational_to_float_systematic_with_naive_impl(::Type{F}, q) where {F}
+    M = Base.MPFR
+    mpfr_rms = (
+        M.MPFRRoundDown, M.MPFRRoundUp, M.MPFRRoundFromZero,
+        M.MPFRRoundToZero, M.MPFRRoundNearest
+    )
+    rms = (
+        RoundDown, RoundFromZero, RoundNearest, RoundToZero, RoundUp,
+        # XXX: these could be tested, but would require more effort to
+        #      test, as MPFR doesn't support them:
+        #
+        # ```julia
+        # RoundNearestTiesAway, RoundNearestTiesUp,
+        # ```
+        ((F <: BigFloat) ? mpfr_rms : ())...
+    )
+    for rm ∈ rms
+        test_rational_to_float_systematic_with_naive_impl(F, q, rm)
+    end
+end
+
+function min_no_promotion_impl(isless::F, a::T, b) where {F, T}
+    if isless(b, a)
+        T(b)::T
+    else
+        a
+    end
+end
+min_no_promotion(a::T, b) where {T} = min_no_promotion_impl(isless, a, b)
+max_no_promotion(a::T, b) where {T} = min_no_promotion_impl(((l, r) -> isless(r, l)), a, b)
+
+safe_typemin(::Type{T}) where {T} = typemin(T)::T
+safe_typemax(::Type{T}) where {T} = typemax(T)::T
+
+safe_typemin(::Type{BigInt}) = BigInt(-10000)
+safe_typemax(::Type{BigInt}) = BigInt( 10000)
```

Comment:
Wouldn't it make sense to use something larger than `typemax(Int128)` here?

Reply:
OK, although I don't think it matters much.